onnxruntime/core/providers/xnnpack/xnnpack_init.h

```diff
@@ -0,0 +1,44 @@
+#include <atomic>
+#include "core/framework/allocator.h"
+#include "core/platform/ort_mutex.h"
+#include "xnnpack.h"
+
+namespace onnxruntime {
+namespace xnnpack {
+
+class XnnpackInitWrapper {
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(XnnpackInitWrapper);
+
+ public:
+  // thread safe
+  static XnnpackInitWrapper& GetInstance() {
+    static XnnpackInitWrapper instance;
+    return instance;
+  }
+
+  void InitXnnpackWithAllocatorAndAddRef(AllocatorPtr allocator);
+  AllocatorPtr GetOrCreateAllocator();
+
+  void increase_ref() {
+    ++allocator_ref_count_;
+  }
+
+  void release_ref() {
+    --allocator_ref_count_;
+    if (allocator_ref_count_ == 0) {
+      ort_allocator_.reset();
+    }
+  }
+
+ private:
+  XnnpackInitWrapper() = default;
+  ~XnnpackInitWrapper() { xnn_deinitialize(); }
+
```

Comment:
if this owns calling xnn_deinitialize it should also own calling xnn_initialize.

Reply:
Refactored with a concise solution

Reply:
To keep its maintainbility, I'd make xnn_init and xnn_deinit to be paired, they have the same calling counts.